setup/Core/SetupOperation.cs

```diff
@@ -1,96 +1,69 @@
+using Terraria.ModLoader.Setup.Core.Utilities;
+
+namespace Terraria.ModLoader.Setup.Core
 {
 	public abstract class SetupOperation
 	{
 		protected delegate void UpdateStatus(string status);
-		protected delegate void Worker(UpdateStatus updateStatus);
+		protected delegate ValueTask Worker(UpdateStatus updateStatus, CancellationToken cancellationToken = default);
 
 		protected class WorkItem
 		{
-			public readonly string status;
-			public readonly Worker worker;
-
 			public WorkItem(string status, Worker worker) {
-				this.status = status;
-				this.worker = worker;
+				Status = status;
+				Worker = worker;
 			}
 
-			public WorkItem(string status, Action action) : this(status, _ => action()) { }
-		}
-
-		protected void ExecuteParallel(List<WorkItem> items, bool resetProgress = true, int maxDegree = 0) {
-			try {
-				if (resetProgress) {
-					taskInterface.SetMaxProgress(items.Count());
-					progress = 0;
-				}
-
-				var working = new List<Ref<string>>();
-				void UpdateStatus() => taskInterface.SetStatus(string.Join("\r\n", working.Select(r => r.item)));
-
-				Parallel.ForEach(Partitioner.Create(items, EnumerablePartitionerOptions.NoBuffering),
-					new ParallelOptions { MaxDegreeOfParallelism = maxDegree > 0 ? maxDegree : Environment.ProcessorCount },
-					item => {
-						taskInterface.CancellationToken.ThrowIfCancellationRequested();
-						var status = new Ref<string>(item.status);
-						lock (working) {
-							working.Add(status);
-							UpdateStatus();
-						}
-
-						void SetStatus(string s) {
-							lock(working) {
-								status.item = s;
-								UpdateStatus();
-							}
-						}
-
-						item.worker(SetStatus);
-
-						lock (working) {
-							working.Remove(status);
-							taskInterface.SetProgress(++progress);
-							UpdateStatus();
-						}
-					});
-			} catch (AggregateException ex) {
-				var actual = ex.Flatten().InnerExceptions.Where(e => !(e is OperationCanceledException));
-				if (!actual.Any())
-					throw new OperationCanceledException();
-
-				throw new AggregateException(actual);
-			}
-		}
+			public WorkItem(string status, Func<CancellationToken, ValueTask> action) : this(status, (_, ct) => action(ct)) { }
 
-		public static string PreparePath(string path)
-			=> path.Replace('/', Path.DirectorySeparatorChar);
+			public WorkItem(string status, Action action) : this(status, (_, _) => { action(); return ValueTask.CompletedTask; }) { }
 
-		public static string RelPath(string basePath, string path) {
-			if (path.Last() == Path.DirectorySeparatorChar)
-				path = path.Substring(0, path.Length - 1);
+			public string Status { get; set; }
 
-			if (basePath.Last() != Path.DirectorySeparatorChar)
-				basePath += Path.DirectorySeparatorChar;
+			public Worker Worker { get; }
+		}
 
-			if (path+Path.DirectorySeparatorChar == basePath) return "";
+		protected async Task ExecuteParallel(
+			List<WorkItem> items,
+			ITaskProgress progress,
+			bool resetProgress = true,
```

Comment:
There are no calls with `resetProgress: false`, so we can remove this parameter.